### PR TITLE
trans: Make sure that closures only get translated once.

### LINKED ISSUE #34569: ICE in debug generation with match guard.
[This](https://github.com/meh/screenruster/tree/rustc-bug) is the full project that causes that failure, make sure to use that commit.

The culprit line is [here](https://github.com/meh/screenruster/blob/rustc-bug/src/locker/mod.rs#L214).

Inside a match:
```rust
e if display.randr.as_ref().map_or(false, |rr| e == rr.event(xrandr::RRScreenChangeNotify))
```

Results in:
```
!dbg attachment points at wrong subprogram for function
!21995 = distinct !DISubprogram(name: "{{closure}}", linkageName: "_ZN12screenruster6locker8{{impl}}5spawn11{{closure}}11{{closure}}E", scope: !21991, file: !352, line: 214, type: !21992, isLocal: true, isDefinition: true, scopeLine: 214, flags: DIFlagPrototyped, isOptimized: false, templateParams: !457, variables: !457)
i1 (i8*, %"locker::display::Extension"*)* @"_ZN12screenruster6locker6Locker5spawn28_$u7b$$u7b$closure$u7d$$u7d$28_$u7b$$u7b$closure$u7d$$u7d$17hb41874d5593ec665E"
  call void @llvm.dbg.declare(metadata %closure.114** %__debuginfo_env_ptr, metadata !59153, metadata !59154), !dbg !59155
!59155 = !DILocation(line: 214, scope: !21990)
!21990 = distinct !DISubprogram(name: "{{closure}}", linkageName: "_ZN12screenruster6locker8{{impl}}5spawn11{{closure}}11{{closure}}E", scope: !21991, file: !352, line: 214, type: !21992, isLocal: true, isDefinition: true, scopeLine: 214, flags: DIFlagPrototyped, isOptimized: false, templateParams: !457, variables: !457)
!21990 = distinct !DISubprogram(name: "{{closure}}", linkageName: "_ZN12screenruster6locker8{{impl}}5spawn11{{closure}}11{{closure}}E", scope: !21991, file: !352, line: 214, type: !21992, isLocal: true, isDefinition: true, scopeLine: 214, flags: DIFlagPrototyped, isOptimized: false, templateParams: !457, variables: !457)
LLVM ERROR: Broken function found, compilation aborted!
```

If I change that to:
```rust
e if e == display.randr.as_ref().map_or(0, |rr| rr.event(xrandr::RRScreenChangeNotify))
```
Then it compiles.

I tried to come up with a reduced test case but failed, so sorry for the full project as test case.

### PR DESCRIPTION
**Pull request by michaelwoerister**
Fixes #34569.

### CODE REVIEW

File: src/librustc_trans/closure.rs [hunk 1]
```diff
@@ -207,41 +207,49 @@ pub fn trans_closure_expr<'a, 'tcx>(dest: Dest<'a, 'tcx>,
...
+                                                              closure_substs);
+        let sig = ty::FnSig {
+            inputs: Some(get_self_type(tcx, closure_def_id, closure_type))
+                        .into_iter().chain(sig.inputs).collect(),
+            output: sig.output,
+            variadic: false
+        };
+
+        trans_closure(ccx,
+                               decl,
```
Comment (luqmana) [inline src/librustc_trans/closure.rs]: Nit: formatting.

**Comment by luqmana** [comment]
LGTM, r=me with nit fixed.

File: src/librustc_trans/closure.rs (hunk 1 above)
Comment (michaelwoerister) [inline src/librustc_trans/closure.rs]: Fixed, thanks!

**Pull request merged**